### PR TITLE
Pixel cookie sharing tests

### LINKED ISSUE #2402: Add tests for pixel cookie sharing heuristic
We probably want:
- A functional test (or tests) for the heuristic itself: https://github.com/EFForg/privacybadger/pull/2147#discussion_r294376018
- Unit tests for `findCommonSubstrings()`: https://github.com/EFForg/privacybadger/pull/2147#discussion_r294376380
- Unit tests for `estimateMaxEntropy`: https://github.com/EFForg/privacybadger/pull/2147#discussion_r294376685

Follows up on #2147.

### PR DESCRIPTION
**Pull request by ablanathtanalba**
Resolves #2402 for both unit and functional testing of the pixel cookie tracking heuristic. 

The unit tests included here covers different aspect of the pixel tracking heuristic to make sure it's catching them and calculating the estimated entropy correctly.

The functional test relies on a resource hosted at gitcdn, which serves a page that recreates the expected functionality of a tracker of this type: it places a cookie on the client browser then generates an img tag that carries a substring of that cookie in the src attribute. The test itself loads that page then checks the snitch map on the headless browser to ensure that that domain has been logged.

Throughout working on this PR, it's been discovered that gitcdn is an unreliable resource for hosting pages needed for these kinds of tests. There are other functional tests in the repository that rely on gitcdn, so it may be best to migrate them all at once to a more reliable resource.

### CODE REVIEW

**Comment by ghostwords** [comment]
What does the testing server look like? We don't have to wait for ops to set it up to review it.

**Comment by ghostwords** [comment]
You can [auto-close issues on GitHub when your pull request is merged by using certain supported keywords in the pull request description](https://help.github.com/en/articles/closing-issues-using-keywords). "Fixes" as one of them; "addresses" is not. You can confirm that the keyword has been recognized by looking for a dashed underline under the word (and a tooltip on hover).

File: src/tests/tests/utils.js [hunk 1]
```diff
@@ -467,4 +467,56 @@ QUnit.test("getHostFromDomainInput", assert => {
...
+    true,
+    "two nearly identical strings of mixed character classes and different cases will return different values"
+  );
+
+  assert.notEqual(
+    utils.estimateMaxEntropy('google.com/analytics') === utils.estimateMaxEntropy('0191/_-google9fjkelo'),
+    true,
+    "strings of the same length but from different character classes will estimate different entropy values"
+  );
+});
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: Missing assertion(s) for the "If all the characters come from one of these common character groups, assume that the group is the domain of possible characters." block. (Right now you can delete that block entirely, and the tests will still pass.)

File: src/tests/tests/utils.js [hunk 2]
```diff
@@ -467,4 +467,56 @@ QUnit.test("getHostFromDomainInput", assert => {
   );
 });
 
+// Tests algorithm used in the pixel tracking heuristic
+// It should return a common substring between two given values
+QUnit.test("findCommonSubstrings", assert => {
+
+  assert.equal(
+    utils.findCommonSubstrings('www.google.com', 'www.google.com/analytics/')[0].length >= 8,
+    true,
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: Too bad there is no "assert greater than" in qUnit ... A bit more direct to use `assert.ok` here then, instead of comparing to `true`.

Comment (ghostwords) [reply src/tests/tests/utils.js]: But I'm not sure this assertion makes sense ... We know the exact length of the common substring for these two strings. But this doesn't prove that all matches have length greater or equal to 8 though.

Should we just test that the substring is exactly what we expect it to be?

File: src/tests/tests/utils.js [hunk 3]
```diff
@@ -467,4 +467,56 @@ QUnit.test("getHostFromDomainInput", assert => {
...
+  assert.equal(
+    utils.findCommonSubstrings('foobar.com/foo/fizz/buzz/bar', 'foobar.com/foo/bizz/fuzz/bar')[0],
+    'foobar.com/foo/',
+    "returns longest matching value from the pair of URLs"
+  );
+
+  assert.equal(
+    utils.findCommonSubstrings('www.google.com', '.w..wgoogle.analytics.com')[0] === undefined,
+    true,
+    "returns empty array if there is no common substring even in very similar string values"
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: Empty array or `undefined`?

File: src/tests/tests/utils.js [hunk 4]
```diff
@@ -467,4 +467,56 @@ QUnit.test("getHostFromDomainInput", assert => {
...
+  );
+});
+
+// used in pixel tracking heuristic, given a string the estimateMaxEntropy function
+// will return the estimated entropy value from it, based on logic parsing the string's length,
+// and classes of character complication included in the string
+QUnit.test("estimateMaxEntropy", assert => {
+  assert.equal(
+    utils.estimateMaxEntropy("google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/analytics.google/google.com/analytics.google/anal") === 257,
+    true,
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: Same as above re ACTUAL and EXPECTED.

File: src/tests/tests/utils.js (hunk 2 above)
Comment (ghostwords) [reply src/tests/tests/utils.js]: We could have another assertion that shows that a common substring under our length threshold is ignored.

File: tests/selenium/pixel_tracking_test.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+class pixelTrackingTesting(pbtest.PBSeleniumTest):
+	"""Tests for the cookie pixel tracking heuristic included in heuristicblocking.js
+		- loads gitcdn resource which places a tracking cookie on page then creates an img tag
+		- img tag makes a src request carrying a substring of that tracking cookie
+		- tracking domain is caught by pixel tracking heuristic, snitch map entry is updated
+	"""
+
+	def test_that_tracker_is_caught(self):
+		TESTING_URL = ("https://gitcdn.link/repo/ablanathtanalba/pixelTrackingTestingResource/master/rsrc.html")
```
Comment (ghostwords) [inline tests/selenium/pixel_tracking_test.py]: Might want to move this test to eff.org/files/ now as gitcdn.link is down again.

Comment (ghostwords) [reply tests/selenium/pixel_tracking_test.py]: The test succeeds even if we remove the contents of the pixel cooking sharing detection function from Privacy Badger prior to running the test. (The test doesn't fail when we need it to fail.) I think this is because the image resource being fetched comes from `github.com`, and it sets cookies. So our regular cookie tracking heuristic catches it.

File: tests/selenium/pixel_tracking_test.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python
+# -*- coding: UTF-8 -*-
+
+import unittest
+import pbtest
+
+class pixelTrackingTesting(pbtest.PBSeleniumTest):
+	"""Tests for the cookie pixel tracking heuristic included in heuristicblocking.js
```
Comment (ghostwords) [inline tests/selenium/pixel_tracking_test.py]: Can you set up your editor to read in our [.editorconfig](https://github.com/EFForg/privacybadger/blob/master/.editorconfig)? Python files should use four spaces instead of tabs.

File: tests/selenium/pixel_tracking_test.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python
+# -*- coding: UTF-8 -*-
+
+import unittest
+import pbtest
+
+class pixelTrackingTesting(pbtest.PBSeleniumTest):
```
Comment (ghostwords) [inline tests/selenium/pixel_tracking_test.py]: Nit: Class names should start with a capital letter (whereas instances start with lower case).

```python
class PixelSharingTest(ptest.PBSeleniumTest):
```

File: tests/selenium/pixel_tracking_test.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+		CHECK_SNITCH_MAP_FOR_ENTRY = (
+				"return chrome.extension.getBackgroundPage()."
+				"badger.storage.snitch_map.getItem('github.com').includes('gitcdn.link');"
+		)
+
+		self.load_url(TESTING_URL)
+		self.load_url(self.options_url)
+
+		# github will already appear in a pretrained badger instance for appearing on fontawesome.com
```
Comment (ghostwords) [inline tests/selenium/pixel_tracking_test.py]: Better to try to guarantee we detected cookie sharing tracking just now (and it wasn't already detected from seed data, nor that we didn't detect some other kind of tracking instead).

You could clear seed data before running the test to eliminate the pre-trained possibility.

To ensure we indeed detected pixel cookie sharing, you could first load a version of the test page that doesn't apply pixel cookie sharing (maybe via a query string flag), verify that tracking was not detected, and then reload the page, this time with pixel cookie sharing enabled, and verify that now tracking was detected.

File: src/tests/tests/utils.js (hunk 3 above)
Comment (ablanathtanalba) [reply src/tests/tests/utils.js]: I decided to test that the first index was undefined as a way of simulating a check that the array would be empty, since if I were to check if two empty arrays were equal to each other it would still run as false. There may be a more graceful way of doing this check

File: src/tests/tests/utils.js (hunk 1 above)
Comment (ablanathtanalba) [reply src/tests/tests/utils.js]: while writing tests for this block I discovered that characters in the `SEP` separator value set aren't ignored like they're supposed to be, causing strings like `google.analytics/123` and `google.ana.lytics/123` to estimate to different entropy values. I confirmed with Bennett that those two test strings should estimate to the same value. I suspect that optimizing the entropy estimating algo is beyond the scope of this PR, so I opened a new issue #2436 to tackle after this PR

File: src/tests/tests/utils.js (hunk 2 above)
Comment (ablanathtanalba) [reply src/tests/tests/utils.js]: Added an assertion to test for substrings under the length threshold are ignored. For the one that was already there, do you think it's overkill to make a randomly generated string of common characters on the fly in this test?

File: tests/selenium/pixel_tracking_test.py (hunk 5 above)
Comment (ablanathtanalba) [reply tests/selenium/pixel_tracking_test.py]: Hmm, I knew that github was already an entry in a pretrained snitch map, which is why I made the check in my test specifically for the inclusion of `gitcdn` in that particular entry, rather than just the existence of the entry itself... Good call on clearing out the pretrained data though, I'll add that in.

File: src/tests/tests/utils.js (hunk 3 above)
Comment (ghostwords) [reply src/tests/tests/utils.js]: `assert.deepEqual` should work ...

File: src/tests/tests/utils.js (hunk 2 above)
Comment (ghostwords) [reply src/tests/tests/utils.js]: Randomness is not a good idea in unit tests generally because it might generate hard-to-reproduce failures.

Is there a code path you are trying to exercise here that isn't exercised by other tests?

File: tests/selenium/pixel_tracking_test.py (hunk 5 above)
Comment (ghostwords) [reply tests/selenium/pixel_tracking_test.py]: Is this in reply to https://github.com/EFForg/privacybadger/pull/2425#discussion_r315423770?

**Comment by ghostwords** [comment]
You should be able to avoid pointless merges like 63a656588e2954a9cbe8fd716b6b56e5cef43fbd (that happen I think when you merge remote into local because your local has commits remote doesn't and you want to push those commits to remote) by first fetching (`git fetch`) the remote and then reapplying your commits over the remote (`git rebase origin/pixel-cookie-sharing-tests`). It's safe to rebase in this case because you aren't rebasing anything you already shared with the world. A bit more information here: https://medium.com/@msanford/rebasing-to-avoid-merge-commits-8f8140375e00.

**Comment by ghostwords** [comment]
I'll merge in master to eliminate the gitcdn.com-caused failures.

File: src/tests/tests/utils.js [hunk 9]
```diff
@@ -477,6 +477,11 @@ QUnit.test("findCommonSubstrings", assert => {
     "returned matches have length greater than or equal to 8"
   );
 
+  assert.notEqual(
+    utils.findCommonSubstrings('www.foo.bar', 'www.foob.ar')[0] === 'www.foo',
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: I think you got the parameters to `assert.notEqual` mixed up here.

Comment (ghostwords) [reply src/tests/tests/utils.js]: It would be more to the point to verify here that there are no results, not that the first result isn't equal to some hypothetical result.

File: tests/selenium/pixel_tracking_test.py (hunk 5 above)
Comment (ablanathtanalba) [reply tests/selenium/pixel_tracking_test.py]: Related to both, yes

Comment (ghostwords) [reply tests/selenium/pixel_tracking_test.py]: I don't understand the connection to this thread.

Comment (ablanathtanalba) [reply tests/selenium/pixel_tracking_test.py]: I misread you and thought that you were saying that `github.com` itself can't be checked for, as it already exists in the pretrained snitch_map, so I was clarifying that that's why I wrote it to explicitly check for the presence of `gitcdn` in the `github` entry. thanks for pointing that out!

Just waiting on resolution from Ops to give me write access, then I'll move my test file over to eff.org/files.

Comment (ghostwords) [reply tests/selenium/pixel_tracking_test.py]: Cool, gotcha. Plus, there is https://github.com/EFForg/privacybadger/pull/2425#discussion_r315421693, which is a larger issue. We need the test to fail when the functionality we mean to test is missing/broken, but the test succeeds because `github.com` sets cookies, and so we detect the `github.com` resource as a tracker but for the wrong reason.

File: src/tests/tests/utils.js (hunk 3 above)
Comment (ghostwords) [reply src/tests/tests/utils.js]: This is a bit of a nit, but a `deepEqual` comparison between the function call and `[]` is more to the point than comparing the index at 0 of the function call to `undefined`. What if there is something at index 1? We want to know that an empty array was returned.

File: src/tests/tests/utils.js [hunk 10]
```diff
@@ -477,20 +477,21 @@ QUnit.test("findCommonSubstrings", assert => {
...
-  assert.notEqual(
-    utils.findCommonSubstrings('www.foo.bar', 'www.foob.ar')[0] === 'www.foo',
+  assert.equal(
+    utils.findCommonSubstrings('www.foo.bar', 'www.foob.ar')[0],
+    undefined,
     "substrings under the length threshold of 8 are ignored"
   );
 
-  assert.equal(
+  assert.deepEqual(
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: Don't need `deepEqual` when comparing strings though.

File: src/tests/tests/utils.js [hunk 11]
```diff
@@ -477,20 +477,21 @@ QUnit.test("findCommonSubstrings", assert => {
     "returned matches have length greater than or equal to 8"
   );
 
-  assert.notEqual(
-    utils.findCommonSubstrings('www.foo.bar', 'www.foob.ar')[0] === 'www.foo',
+  assert.equal(
```
Comment (ghostwords) [inline src/tests/tests/utils.js]: Same comment as https://github.com/EFForg/privacybadger/pull/2425/commits/7555cfc30cdee9457091b3a65c92dcdc6d2b4370#r316766940

File: tests/selenium/pixel_tracking_test.py [hunk 12]
```diff
@@ -14,11 +14,17 @@ class PixelTrackingTesting(pbtest.PBSeleniumTest):
 	def test_that_tracker_is_caught(self):
 		TESTING_URL = ("https://gitcdn.link/repo/ablanathtanalba/pixelTrackingTestingResource/master/rsrc.html")
 
+		CLEAR_TRAINED_DATA = (
+		    "chrome.extension.getBackgroundPage().badger.storage.snitch_map.updateObject({})"
```
Comment (ghostwords) [inline tests/selenium/pixel_tracking_test.py]: This is OK, but we should probably reuse an existing method (`badger.storage.clearTrackerData()`): https://github.com/EFForg/privacybadger/blob/b52f23d5c9ba03811de9aaad67aa60e1b84292d5/tests/selenium/options_test.py#L54

Comment (ghostwords) [reply tests/selenium/pixel_tracking_test.py]: I think you need to visit the options page first, before running this JavaScript.

Comment (ablanathtanalba) [reply tests/selenium/pixel_tracking_test.py]: Makes sense, I'll flip that over

File: tests/selenium/pixel_tracking_test.py (hunk 8 above)
Comment (ablanathtanalba) [reply tests/selenium/pixel_tracking_test.py]: Cleared up this and [this](https://github.com/EFForg/privacybadger/pull/2425#discussion_r316770343) in my most recent local commit, including the proposed URLs for the testing resources when I get write access to the server.
